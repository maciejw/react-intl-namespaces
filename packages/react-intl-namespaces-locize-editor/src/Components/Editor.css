body.searchEnabled * {
  pointer-events: auto;
}
