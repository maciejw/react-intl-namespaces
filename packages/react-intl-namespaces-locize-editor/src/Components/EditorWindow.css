.container {
  bottom: 0;
  box-shadow: -3px 0 5px 0 rgba(0, 0, 0, 0.5);
  position: fixed;
  right: 0;
  top: 0;
  z-index: 2000;
  display: none;
}
.containerVisible {
  display: block;
}

.iframe {
  border: none;
  height: 100%;
}
