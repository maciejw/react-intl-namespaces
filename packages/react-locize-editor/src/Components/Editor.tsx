import * as React from 'react';
import { LocizeEditorBinding } from '../locizeEditorBinding';

const editor = new LocizeEditorBinding({
  backend: {
    apiKey: '6b9ac145-b395-47dc-bcc4-001398c4c843',
    projectId: '2592abb8-1129-457d-a06b-836745d33c55',
    referenceLng: 'en',
  },
  enabled: true,
});

export class Editor extends React.Component {
  constructor(props: {}, context: {}) {
    super(props, context);
  }
  public render() {
    return null;
  }
}
