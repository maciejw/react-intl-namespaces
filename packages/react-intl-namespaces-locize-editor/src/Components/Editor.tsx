import * as React from 'react';
import * as ReactDOM from 'react-dom';

import { invariant } from '../invariant';
import { EditorPanel } from './EditorPanel';
import { EditorWindow } from './EditorWindow';

export class Editor extends React.Component<
  Editor.RequiredProps & Partial<Editor.OptionalProps>
> {
  public render() {
    return <EditorComponent {...Editor.defaultProps} {...this.props} />;
  }
}
class EditorComponent extends React.Component<
  Editor.RequiredProps & Editor.OptionalProps,
  Editor.State
> {
  private message: (ev: MessageEvent) => void;
  // prettier-ignore
  private keypress: (ev: KeyboardEvent) => void;
  // prettier-ignore
  private click: (e: MouseEvent) => void;
  // prettier-ignore
  private postMessage!: Promise<EditorWindow.PostMessage>;
  // prettier-ignore

  private editor: HTMLElement;

  constructor(props: Editor.RequiredProps & Editor.OptionalProps, context: {}) {
    super(props, context);
    this.state = {
      searchEnabled: true,
      showEditor: this.props.enabled,
      showIds: false,
    };

    let editor = document.getElementById('locize-editor');
    // invariant(editor === null, 'You should use only one Editor');

    editor = document.createElement('div');
    editor.id = 'locize-editor';
    document.body.appendChild(editor);
    this.editor = editor;

    this.click = (ev: MouseEvent) => this.lookupInEditor(ev);

    const { toggleKeyModifier, toggleKeyCode } = this.props;

    this.keypress = (ev: KeyboardEvent) => {
      if (ev[toggleKeyModifier] && ev.which === toggleKeyCode) {
        this.onSearchEnabled();
      }
    };
    this.message = (ev: MessageEvent) => {
      if (ev.data[toggleKeyModifier] && ev.data.which === toggleKeyCode) {
        this.onSearchEnabled();
      }
    };

    document.body.addEventListener('click', this.click);
    document.addEventListener('keypress', this.keypress);
    window.addEventListener('message', this.message);
  }

  public componentWillUnmount() {
    document.body.removeChild(this.editor);
    document.removeEventListener('keypress', this.keypress);
    window.removeEventListener('message', this.message);
  }

  public render() {
    const { searchEnabled, showIds } = this.state;
    return ReactDOM.createPortal(
      <div data-ignore-locize-editor="true">
        <EditorPanel
          showIds={showIds}
          searchEnabled={searchEnabled}
          onRefresh={() => this.onRefresh()}
          onSearchEnabled={() => this.onSearchEnabled()}
          onShowIds={() => this.onShowIds()}
          language={this.props.language}
          onChangeLanguage={this.props.onChangeLanguage}
          getLanguages={this.props.getLanguages}
        />
        <EditorWindow
          mode={this.props.mode}
          editorWidthInPixels={this.props.editorWidthInPixels}
          url={this.props.url}
          onOpen={i => this.open(i)}
        />
      </div>,
      this.editor,
    );
  }
  private onShowIds() {
    const { showIds: oldShowIds, ...rest } = this.state;
    const showIds = !oldShowIds;
    this.setState({ showIds, ...rest });
    this.props.onShowIds(showIds);
  }
  private open(callback: Promise<EditorWindow.PostMessage>) {
    this.postMessage = callback;
  }

  private isTranslatedOrIgnored(element: Element) {
    if (this.isHtmlElement(element)) {
      let e = element;
      while (e.tagName.toLowerCase() !== 'body') {
        if (
          e.dataset.ignoreLocizeEditor === 'true' ||
          e.dataset.translated === 'true'
        ) {
          return true;
        }
        if (e.parentElement === null) {
          return false;
        }
        e = e.parentElement;
      }
    }
    return false;
  }
  private isHtmlElement(element: Element): element is HTMLElement {
    return element instanceof HTMLElement;
  }

  private async lookupInEditor(e: MouseEvent) {
    if (!this.state.searchEnabled) {
      return;
    }
    e.preventDefault();

    if (e.srcElement) {
      const resourceContainer: HTMLElement | null = e.srcElement.parentElement;

      if (!resourceContainer || this.isTranslatedOrIgnored(resourceContainer)) {
        return;
      }

      const {
        ns,
        key,
        defaultMessage,
        description,
      } = resourceContainer.dataset;

      let message: Editor.SearchMessage;

      if (ns && key) {
        message = this.createMessage(ns, key, defaultMessage, description);

        const postMessage = await this.postMessage;

        if (this.props.mode === 'window') {
          if (!window.closed) {
            postMessage(message, this.props.url);
          }
        }
        if (this.props.mode === 'iframe') {
          postMessage(message, this.props.url);
        }
      } else {
        alert(
          `Missing key and namespace of resource, try search for a text or select ID mode`,
        );
      }
    }
  }
  private createMessage(
    ns: string,
    key: string,
    defaultMessage?: string,
    description?: string,
  ): Editor.SearchMessage {
    const { language: lng, projectId, version } = this.props;
    return {
      lng,
      message: 'searchForKey',
      ns,
      projectId,
      token: key,
      version,
    };
  }

  private onSearchEnabled() {
    const { searchEnabled } = this.state;
    this.setState({ searchEnabled: !searchEnabled });
  }
  private onRefresh() {
    this.props.onRefresh();
  }
}

export namespace Editor {
  export interface State {
    searchEnabled: boolean;
    showIds: boolean;
    showEditor: boolean;
  }
  export const defaultProps: Editor.OptionalProps = {
    editorWidthInPixels: 700,
    enabled: false,
    getLanguages: () => [],
    mode: 'iframe',
    onChangeLanguage: language => void 0,
    onRefresh: () => void 0,
    onShowIds: show => void 0,
    toggleKeyCode: 24,
    toggleKeyModifier: 'ctrlKey',
    url: 'https://www.locize.io',
    version: 'latest',
  };

  export interface OptionalProps {
    editorWidthInPixels: number;
    enabled: boolean;
    toggleKeyCode: number;
    toggleKeyModifier: 'ctrlKey' | 'altKey' | 'shiftKey';
    mode: 'window' | 'iframe';
    url: string;
    version: string;
    onShowIds: (show: boolean) => void;
    onChangeLanguage: (language: string) => void;
    getLanguages: () => string[];
    onRefresh: () => void;
  }

  export interface RequiredProps {
    apiKey: string;
    projectId: string;
    referenceLanguage: string;
    language: string;
  }
  export interface SearchMessage {
    message: 'searchForKey';
    projectId: string;
    version: string;
    lng: string;
    ns: string;
    token: string;
    defaultMessage?: string;
    description?: string;
  }
}
