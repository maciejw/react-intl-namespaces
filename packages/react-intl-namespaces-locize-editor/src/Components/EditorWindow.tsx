import * as React from 'react';
import * as ReactDOM from 'react-dom';

import * as styles from './EditorWindow.css';

function delay(timeout: number = 0) {
  return new Promise<void>(resolve => {
    window.setTimeout(resolve, timeout);
  });
}
export class EditorWindow extends React.Component<EditorWindow.Props> {
  // prettier-ignore
  private window: Window | undefined;
  constructor(props: EditorWindow.Props, state: {}) {
    super(props, state);
  }

  public componentDidMount() {
    if (this.props.mode === 'window') {
      const openedWindow = window.open(
        this.props.url,
        'locize-editor',
        '',
        true,
      );

      if (openedWindow !== null) {
        this.props.onOpen(
          new Promise(async resolve => {
            await delay(3000);
            resolve(openedWindow);
          }),
        );
      }
    }
  }

  public render() {
    switch (this.props.mode) {
      case 'iframe':
        const { container, iframe } = EditorWindow.inlineStyles(
          this.props.editorWidthInPixels,
        );
        return (
          <div
            className={styles.container}
            style={container}
            data-ignore-locize-editor="true"
          >
            <iframe
              className={styles.iframe}
              ref={e => {
                if (e !== null) {
                  this.props.onOpen(
                    new Promise(resolve => {
                      resolve(e.contentWindow);
                    }),
                  );
                }
              }}
              style={iframe}
              src={this.props.url}
            />
          </div>
        );
      case 'window':
        return null;
    }
  }
}
export namespace EditorWindow {
  export interface Props {
    url: string;
    onOpen: (window: Promise<Window>) => void;
    editorWidthInPixels: number;
    mode: 'iframe' | 'window';
  }

  export const inlineStyles: (
    editorWidthInPixels: number,
  ) => Record<
    'container' | 'iframe',
    React.CSSProperties
  > = editorWidthInPixels => ({
    container: {
      width: `${editorWidthInPixels}px`,
    },
    iframe: {
      width: `${editorWidthInPixels}px`,
    },
  });
}
