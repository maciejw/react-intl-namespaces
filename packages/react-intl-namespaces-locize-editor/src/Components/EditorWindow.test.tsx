import { mount } from 'enzyme';
import * as React from 'react';
import { Editor } from './Editor';
import { EditorWindow } from './EditorWindow';

function delay(timeout: number = 0) {
  return new Promise<void>(resolve => {
    window.setTimeout(resolve, timeout);
  });
}

const options: Editor.RequiredProps = {
  apiKey: '6b9ac145-b395-47dc-bcc4-001398c4c843',
  language: 'en',
  projectId: '2592abb8-1129-457d-a06b-836745d33c55',
  referenceLanguage: 'en',
};
describe('EditorWindow', () => {
  it('should show panel in iframe mode', () => {
    const getLanguagesMock = jest.fn();

    const PropsMock = jest.fn<EditorWindow.Props>(() => ({
      onOpen: jest.fn(),
    }));

    const mock = new PropsMock();

    const wrapper = mount(
      <EditorWindow editorWidthInPixels={100} mode="iframe" {...mock} />,
    );
    expect(wrapper.html()).toMatchSnapshot();

    expect(mock).toMatchSnapshot();
  });

  it('should show panel in window mode', async done => {
    const getLanguagesMock = jest.fn();

    const PropsMock = jest.fn<EditorWindow.Props>(() => ({
      onOpen: jest.fn(),
    }));

    const mock = new PropsMock();

    const wrapper = mount(
      <EditorWindow editorWidthInPixels={100} mode="window" {...mock} />,
    );

    await delay(3030);
    expect(wrapper.html()).toMatchSnapshot();

    expect(mock).toMatchSnapshot();
    done();
  });
});
