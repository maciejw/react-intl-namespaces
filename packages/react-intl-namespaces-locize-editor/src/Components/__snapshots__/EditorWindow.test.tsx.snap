// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorWindow should show panel in iframe mode 1`] = `"<div class=\\"container\\" style=\\"width: 100px;\\" data-ignore-locize-editor=\\"true\\"><iframe class=\\"iframe\\" style=\\"width: 100px;\\"></iframe></div>"`;

exports[`EditorWindow should show panel in iframe mode 2`] = `
Object {
  "onOpen": [MockFunction] {
    "calls": Array [
      Array [
        Promise {},
      ],
    ],
  },
}
`;

exports[`EditorWindow should show panel in window mode 1`] = `null`;

exports[`EditorWindow should show panel in window mode 2`] = `
Object {
  "onOpen": [MockFunction] {
    "calls": Array [
      Array [
        Promise {},
      ],
    ],
  },
}
`;
